### PR TITLE
PEPPER-18: Adjust AOM checks

### PR DESCRIPTION
**Pull request by Trimax**
PEPPER-18
--

This PR changes the validation of AOM from a simple comparison to **18** to a more tricky one:
aom

In other words, the validations are:

**Child**

- ENROLLING_CHILD.AGE >= 7 && !ADD_PARTICIPANT_INCAPACITATED && CONSENT.AGE <= 7
- ENROLLING_CHILD.AGE < 7 && CONSENT.AGE >= 7 && CONSENT.AGE < 18
- CONSENT_PARENTAL_CHILD_DATE_OF_BIRTH >= AOM
- ENROLLING_CHILD_AGE >= AOM

**Dependent**

- CONSENT_DEPENDENT_DATE_OF_BIRTH >= AOM && CONSENT.AGE <= AOM
- ENROLLING_DEPENDENT_AGE >= AOM && !ADD_PARTICIPANT_INCAPACITATED_DEPENDENT
- ENROLLING_DEPENDENT_AGE < AOM

**Self**

- CONSENT.CONSENT_SELF_DATE_OF_BIRTH < AOM

Once merged, **SingularAgeValidationUpdate** task has to be executed

### Setup
1. Create a new DSS database, and deploy the `singular` study
2. Run `studybuilder-cli` task `SingularAOMAddParticipantUpdates`
3. Run `studybuilder-cli` task `SingularAgeValidationUpdate`
4. Start the `cmi-singular` Angular client & DSS locally.
5. Enroll a new user, and walk through the AoM diagrams.

### CODE REVIEW

**Comment by ssettipalli** [comment]
Validation pex looks ok but 
I dont see a patch to update/deploy the AOM validation pex expressions ?

**Review by ssettipalli — CHANGES_REQUESTED** [review]
We need a patch to deploy the AOM validation PEX expr changes I believe.

**Comment by Trimax** [comment]
> Validation pex looks ok but I dont see a patch to update/deploy the AOM validation pex expressions ?

The files I modified are being used by the patch that I wrote before. It updates/recreates the data accordingly.

**Review by ssettipalli — DISMISSED** [review]

**Comment by bskinner** [comment]
My local tests of this failed- I attempted to enroll a child in Alberta, Canada at the age of 18 and no errors were presented. I'm looking into the issue now.

File: pepper-apis/studybuilder-cli/studies/singular/patches/ddp-8554-parental-validations.conf [hunk 1]
```diff
@@ -68,18 +68,74 @@
...
         ]
       },
-      "stableIds": ["CONSENT_PARENTAL_CHILD_DATE_OF_BIRTH", "ENROLLING_CHILD_AGE"],
+      "stableIds": ["CONSENT_PARENTAL_CHILD_DATE_OF_BIRTH", "ADD_PARTICIPANT_COUNTRY_CHILD", "ADD_PARTICIPANT_STATE_CHILD", "ADD_PARTICIPANT_PROVINCE_CHILD", "ENROLLING_CHILD_AGE"],
       "precondition": """
                           user.studies["singular"].forms["ADD_PARTICIPANT_PARENTAL"].questions["ENROLLING_CHILD_AGE"].isAnswered()
                           &&
+                          user.studies["singular"].forms["ADD_PARTICIPANT_PARENTAL"].questions["ADD_PARTICIPANT_COUNTRY_CHILD"].isAnswered()
+                          &&
                           user.studies["singular"].forms["CONSENT_PARENTAL"].questions["CONSENT_PARENTAL_CHILD_DATE_OF_BIRTH"].isAnswered()
```
Comment (bskinner) [inline pepper-apis/studybuilder-cli/studies/singular/patches/ddp-8554-parental-validations.conf:77]: This question is not guaranteed to be answered when the user is filling out the `ADD_PARTICIPANT_PARENTAL` form. This results in the AoM conditions not being evaluated or properly applied as the precondition evaluates to `false`.

**Review by bskinner — CHANGES_REQUESTED** [review]

File: pepper-apis/studybuilder-cli/studies/singular/patches/ddp-8554-parental-validations.conf (hunk 1 above)
Comment (Trimax) [reply pepper-apis/studybuilder-cli/studies/singular/patches/ddp-8554-parental-validations.conf:77]: Now I know what is this field for. Simplified :)

**Review by bskinner — CHANGES_REQUESTED** [review]
[Line 82](https://github.com/broadinstitute/ddp-study-server/blob/5b2c4fa4e837439381961a584061cc6995790717/pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAgeValidationUpdate.java#L82) in `SingularAgeValidationUpdate.java` is breaking existing behavior for the activities modified in this patch- it is removing all existing validations before adding the new ones.

**Comment by Trimax** [comment]
After Friday's discussion patch was reworked, we're not allowed to modify the original `study-activities.conf` file, so the patch is responsible for cleaning up existing validations. A new version of the patch finds validations that match to the precondition and expression specified and removes them.

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAgeValidationUpdate.java [hunk 2]
```diff
@@ -61,30 +66,59 @@ private void updateValidations(final Handle handle, final ActivityBuilder builde
...
-        @SqlUpdate("DELETE FROM activity_validation WHERE study_activity_id = :studyActivityId")
-        void deleteValidations(@Bind("studyActivityId") long studyActivityId);
+        @SqlUpdate("DELETE av "
+                 + "FROM activity_validation av "
+                 + "JOIN study_activity sa ON av.study_activity_id = sa.study_activity_id "
+                 + "WHERE sa.study_activity_code = :activityCode "
+                 + "AND REPLACE(REPLACE(REPLACE(REPLACE(av.expression_text, ' ', ''), CHAR(13), ''), CHAR(10), ''), CHAR(9), '') = "
+                 + "    REPLACE(REPLACE(REPLACE(REPLACE(:expression, ' ', ''), CHAR(13), ''), CHAR(10), ''), CHAR(9), '') "
+                 + "AND REPLACE(REPLACE(REPLACE(REPLACE(av.precondition_text, ' ', ''), CHAR(13), ''), CHAR(10), ''), CHAR(9), '') = "
+                 + "    REPLACE(REPLACE(REPLACE(REPLACE(:precondition, ' ', ''), CHAR(13), ''), CHAR(10), ''), CHAR(9), '') ")
```
Comment (bskinner) [inline pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAgeValidationUpdate.java:125]: Can you add a comment to this method detailing why we're performing the string of `REPLACE` calls?

**Review by bskinner — COMMENTED** [review]
A minor comment to add, but looks good otherwise!

File: pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAgeValidationUpdate.java (hunk 2 above)
Comment (Trimax) [reply pepper-apis/studybuilder-cli/src/main/java/org/broadinstitute/ddp/studybuilder/task/SingularAgeValidationUpdate.java:125]: Added! I hope the explanation is clear enough :)

**Review by bskinner — APPROVED** [review]
👍

**Review by MocanaAtBroad — APPROVED** [review]

**Pull request merged**